File: solutions/350-intersection-of-two-arrays-ii/intersection-of-two-arrays-ii.ts
function intersect(nums1: number[], nums2: number[]): number[] {

    const freqMap: Record<number, number> = {};
    for(let i = 0; i < nums1.length; i += 1) {
        if(!freqMap.hasOwnProperty(nums1[i])) {
            freqMap[nums1[i]] = 0;
        }
        freqMap[nums1[i]] += 1;
    }

    const res = [];
    for(let i = 0; i < nums2.length; i += 1) {
        if(freqMap.hasOwnProperty(nums2[i])) {
            res.push(nums2[i]);
            freqMap[nums2[i]] -= 1;
            if(freqMap[nums2[i]] === 0) {
                delete freqMap[nums2[i]];
            }
        }
    }
    return res;
};

File: solutions/350-intersection-of-two-arrays-ii/intersection-of-two-arrays-ii.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        
        result = []
        num_count = {}

        for num in nums1:
            if num not in num_count:
                num_count[num] = 0
            num_count[num] += 1

        for num in nums2:
            if num in num_count:
                result.append(num)
                num_count[num] -= 1
                if num_count[num] == 0:
                    del num_count[num]
        
        return result